utils/toc2cue.1
.TH "TOC2CUE" "1" "2006-11-17" "1.2.2" "CDRDAO"

.SH "NAME"
toc2cue \- converts a TOC files of
.I cdrdao(1)
into a .cue file

.SH "SYNOPSIS"
.B toc2cue
.BR [\| \-v
.IB verbosity_level \|]
.I input_toc_file output_cue_file
.PP
.B toc2cue
.BR [\| \-V \|]

.SH "DESCRIPTION"
.B toc2cue
converts a TOC file of
.I cdrdao(1)
into a .cue file.

Please note that the resulting cue file is only valid if the toc-file was created with
.I cdrdao(1)
using the commands 'read-toc' or 'read-cd'. For manually created or edited toc-files the cue file may not be correct. This program just checks for the most obvious toc-file features that cannot be converted to a cue file. Furthermore, if the toc-file contains audio tracks the byte order of the image file will be wrong. This will result in static noise when the cue file is used for recording (even with
.I cdrdao(1)
itself).

.SH "OPTIONS"
.TP
.BI "\-v " verbosity_level
Sets the verbosity level to use, can be a integer between 1 and 3. By default, only fatal errors are displayed. A verbosity level of 1 shows warnings, 2 warnings and errors, 3 warnings, errors and internal errors.
.TP
.B \-V
Prints the version of toc2cue.

.SH "SEE ALSO"
.I cdrdao(1)

.SH "AUTHOR"
.B toc2cue
was written by Andreas Mueller <andreas@daneb.de> and is part of
.I cdrdao(1).
.PP
This manual page was written by Francois Wendling <frwendling@free.fr> and revised by Daniel Baumann <daniel@debian.org>, for the Debian project (but may be used by others).


ChangeLog
--------------------------------------------------------------------------
  cdrdao 1.2.3
--------------------------------------------------------------------------

 o  By default, cdrdao tries to build with a native SCSI interface. Use
    "--with-libscg" to use bundled libscg library. (Denis Leroy)

 o  Complete rewrite of native Linux SG driver for SG 3.0 using SG_IO
    ioctl. That should make it behave better with ConsoleKit based
    permissions environment. Cdrdao should now be compiled on Linux
    without cdrtools by using "--without-scglib". (Denis Leroy)

 o  Added support for new FIRST_TRACK_NO keyword to allow a CD to start
    with a track number other than 1. (Edgar Fuss)

 o  Added native interfaces for Irix, Mac OSX and NetBSD. (Edgar Fuss)

 o  Fixed bug in Toc file parser track length computation. (Denis Leroy)

 o  Man page typo fixes. (Denis Leroy)

 o  Cygwin compilation fixes. Other various compilation warning
    fixes. (Denis Leroy)

 o  Some scsilib fixes for FreeBSD, 32-bit and 64-bit SPARC. (Marius
    Strobl)

 o  Some makefile changes to avoid conflict with installed system libs.
    (Marius Strobl)

 o  Added patch to fix ARM and MIPS scsilib compile (Kedar Sovani, Debian)

 o  Added man pages for toc2cue and toc2cddb, from Debian.

 o  Move privilege drop earlier in main flow (based on Debian patch)

GCDMaster changes:

 o  Preferences now stored in GConf, exports schema file (Denis Leroy)

 o  Added simple Preferences dialog to change temporary file directory
    (Denis Leroy)

 o  Fixed GCC 4.3 compilation issues (Denis Leroy)

--------------------------------------------------------------------------
  cdrdao 1.2.2 (2006-09-20)
--------------------------------------------------------------------------

 o  Added device locking for win32 (Giuseppe Corbelli)

 o  Man page cleanup (Andrew Suffield)

GCDMaster changes:

 o  Fixed GCC 4.1 compilation issues

 o  Fixed hang when inserting WAV file


--------------------------------------------------------------------------
  cdrdao 1.2.1 (2005-10-28)
--------------------------------------------------------------------------

 o  Fixed bad bug with missing return statemements in recomputeLength
    and format conversion.

 o  Fixed broken '-' STDIN input feature.

 o  Fixed various gcc4-related compile problems.

 o  Fixed copy problem with single device when read_device is
    specified.

 o  Added autodetect of cdrtools header files in configure

 o  Minor fixes to driver table.

 o  pccts build is now better integrated with automake/autoconf.

 o  Added FreeBSD 7 support to configure.ac. Fixed libao dependency.	

GCDMaster changes:

 o  The crash when creating a new "Create Audio CD" view seemed to be
    related to gtkmm and libsigc++ and went away with newer releases of
    those packages.

toc2cue changes:

 o  toc2cue: fixed bug that was causing empty output files.

 o  Added conversion of COPY to DCP flags. (Julio Sanchez Fernandez)


--------------------------------------------------------------------------
  cdrdao 1.2.0 (2005-05-12)
--------------------------------------------------------------------------

 o  SECURITY FIX: cdrdao now gives up its root privileges after setting
    up real-time scheduling, as well as before saving settings through
    the --save option. This fixes a potential local root exploit when
    cdrdao is installed with the +s chmod flag. Using --save now also
    forces an early exit after the settings are saved.

 o  Added MP3 and Ogg Vorbis file support (through respectively the
    libmad and libvorbis libraries). cdrdao will decode the MP3 and
    Ogg files into temporary WAV files that will be deleted upon exit
    (unless the new option --keep is used). The directory used to
    store those temporary WAV files can be specified with the --tmpdir
    option (default is /tmp).

 o  Improved native CUE file support: replaced old incomplete existing
    parser with the one from Matthias Czapla's excellent
    cue2toc. Added support for cutting binary files.

 o  Added --rspeed option to manually set the device reading speed. Be
    warned not all devices support this.

 o  Packaged scsilib library upgraded from cdrtools 2.01 (previously
    was from 2.01a31).

 o  Added --no-mode2-mixed option. Don't read a MODE2 disk as
    MODE2_FORM_MIXED, only read MODE2_FORM1 or MODE2_FORM2 (Matthieu
    Castet).

 o  Added help for little-known drive-info command.

GCDMaster changes:

 o  MP3 and Ogg Vorbis support: you can drag and drop .mp3, .m3u and
    .ogg files from Nautilus into the sample display window.

 o  Switched to gtkmm24 API. Improved file browsers.

 o  CUE files support.

 o  Sound output now uses libao library.

 o  Added 'Select All' item in menu.

 o  Added 'Eject' button to progress dialog

 o  Bug fixes (sample selection weirdness, couldn't close window during
    play, problems with gcdmaster command-line argument, crashes with
    multiple project windows, drive status not reported correctly).

--------------------------------------------------------------------------
  cdrdao 1.1.9 (2004-06-07)
--------------------------------------------------------------------------

 o MMC-driver auto-detection. If no --driver option is specified,
   cdrdao will try to autodetect an MMC-compliant device and use the
   generic-mmc driver by default. (Denis Leroy)

 o Included scsilib upgraded from cdrtools 2.01a31 (Denis Leroy)

 o Added --eject option to the unlock command to eject the cd after
   unlocking. (Eric Benson)

 o New 'discid' command to display CDDB information. (Eric Benson,
   Denis Leroy).

 o Fixed scsilib compiling problems with Linux kernel 2.6 (Jim Gifford)

 o Fixed compiling problems with gcc 3.4 (Jim Gifford, Denis Leroy)

 o Added --full-burn and --capacity options to extend lead-out to entire disc,
   beyond 80 mins limit. (Vladimir Petrov)


GCDMaster changes (Denis Leroy):

 o Device configuration dialog now scans for devices on both ATA and

utils/toc2cue.cc
/*  toc2cue - converts cdrdao's toc-files to .cue files
 *
 *  Copyright (C) 2001 Andreas Mueller <andreas@daneb.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <stdarg.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <fstream>

#include "util.h"
#include "Toc.h"

static const char *PRGNAME = NULL;
static int VERBOSE = 1;

void message(int level, const char *fmt, ...)
{
  long len = strlen(fmt);
  char last = len > 0 ? fmt[len - 1] : 0;

  va_list args;
  va_start(args, fmt);

  if (level < 0) {
    switch (level) {
    case -1:
      fprintf(stderr, "WARNING: ");
      break;
    case -2:
      fprintf(stderr, "ERROR: ");
      break;
    case -3:
      fprintf(stderr, "INTERNAL ERROR: ");
      break;
    default:
      fprintf(stderr, "FATAL ERROR: ");
      break;
    }
    vfprintf(stderr, fmt, args);
    if (last != ' ' && last != '\r')
      fprintf(stderr, "\n");
    
    fflush(stderr);
    if (level <= -10)
      exit(1);
  }
  else if (level <= VERBOSE) {
    vfprintf(stderr, fmt, args);
    if (last != ' ' && last != '\r')
      fprintf(stderr, "\n");

    fflush(stderr);
  }

  va_end(args);
}

static void printVersion()
{
  message(1, "toc2cue version %s - (C) Andreas Mueller <andreas@daneb.de>",
	  VERSION);
  message(1, "");
}

static void printUsage()
{
  message(0, "\nUsage: %s [-v #] { -V | input-toc-file output-cue-file}", PRGNAME);
}

static int parseCommandLine(int argc, char **argv, char **tocFile,
			    char **cueFile)
{
  int c;
  int printVersion = 0;
  extern char *optarg;
  extern int optind, opterr, optopt;

  opterr = 0;

  while ((c = getopt(argc, argv, "Vv:")) != EOF) {
    switch (c) {
    case 'V':
      printVersion = 1;
      break;

    case 'v':
      if (optarg != NULL) {
	if ((VERBOSE = atoi(optarg)) < 0) {
	  message(-2, "Invalid verbose level: %s", optarg);
	  return 0;
	}
      }
      else {
	message(-2, "Missing verbose level after option '-v'.");
	return 0;
      }
      break;

    case '?':
      message(-2, "Invalid option: %c", optopt);
      return 0;
      break;
    }
  }

  if (printVersion) {
    return 1;
  }

  if (optind < argc) {
    *tocFile = strdupCC(argv[optind]);
    optind++;
  }
  else {
    message(-2, "Missing toc-file name.");
    return 0;
  }

  if (optind < argc) {
    *cueFile = strdupCC(argv[optind]);
    optind++;
  }
  else {
    message(-2, "Missing cue file name.");
    return 0;
  }

  if (optind != argc) {
    message(-2, "Expecting exactly two arguments.");
    return 0;
  }

  return 2;
}

int main(int argc, char **argv)
{
  char *tocFile, *cueFile;
  Toc *toc;

  PRGNAME = *argv;
  
  switch (parseCommandLine(argc, argv, &tocFile, &cueFile)) {
  case 0:
    printUsage();
    return 1;
    break;
    
  case 1:
    printf("%s\n", VERSION);
    return 0;
    break;
  }

  printVersion();

  if ((toc = Toc::read(tocFile)) == NULL) {
    message(-2, "Failed to read toc-file '%s'.", tocFile);
    return 1;
  }

  Msf start, end;
  const Track *trun;
  int trackNr;
  TrackIterator titr(toc);
  char *binFileName = NULL;
  int err = 0;

  // first make some consistency checks, surely not complete to identify
  // toc-files that can be correctly converted to cue files
  for (trun = titr.first(start, end), trackNr = 1;
       trun != NULL;
       trun = titr.next(start, end), trackNr++) {
    const SubTrack *strun;
    int stcount;
    TrackData::Type sttype1 = TrackData::DATAFILE,
        sttype2 = TrackData::DATAFILE;
    SubTrackIterator stitr(trun);

    switch (trun->type()) {
    case TrackData::MODE0:
    case TrackData::MODE2_FORM2:
      message(-2, "Cannot convert: track %d has unsupported mode.", trackNr);
      err = 1;
      break;
    default:
      break;
    }

    for (strun = stitr.first(), stcount = 0;
	 strun != NULL;
	 strun = stitr.next(), stcount++) {

      // store types of first two sub-tracks for later evaluation
      switch (stcount) {
      case 0:
	sttype1 = strun->TrackData::type();
	break;
      case 1:
	sttype2 = strun->TrackData::type();
	break;
      }

      // check if whole toc-file just references a single bin file
      if (strun->TrackData::type() == TrackData::DATAFILE) {
	if (binFileName == NULL) {
	  binFileName = strdupCC(strun->filename());
	}
	else {
	  if (strcmp(binFileName, strun->filename()) != 0) {
	    message(-2, "Cannot convert: toc-file references multiple data files.");
	    err = 1;
	  }
	}
      }
    }

    switch (stcount) {
    case 0:
      message(-2, "Cannot convert: track %d references no data file.",
	      trackNr);
      err = 1;
      break;

    case 1:
      if (sttype1 != TrackData::DATAFILE) {
	message(-2, "Cannot convert: track %d references no data file.",
	      trackNr);
	err = 1;
      }
      break;

    case 2:
      if (sttype1 != TrackData::ZERODATA || sttype2 != TrackData::DATAFILE) {
	message(-2, "Cannot convert: track %d has unsupported layout.",
		trackNr);
	err = 1;
      }
      break;

    default:
      message(-2, "Cannot convert: track %d has unsupported layout.", trackNr);
      err = 1;
      break;
    }
  }

  if (binFileName == NULL) {
    message(-2, "Cannot convert: toc-file references no data file.");
    err = 1;
  }

  if (err) {
    message(-2, "Cannot convert toc-file '%s' to a cue file.", tocFile);
    return 1;
  }

  std::ofstream out(cueFile);

  if (!out) {
    message(-2, "Cannot open cue file \'%s\' for writing: %s", cueFile,
	    strerror(errno));
    return 1;
  }

  out << "FILE \"" << binFileName << "\" BINARY" << "\n";

  long offset = 0;

  for (trun = titr.first(start, end), trackNr = 1;
       trun != NULL;
       trun = titr.next(start, end), trackNr++) {
    char buf[20];

    sprintf(buf, "%02d ", trackNr);
    out << "  TRACK " << buf;

    switch (trun->type()) {
    case TrackData::AUDIO:
      out << "AUDIO";
      break;
    case TrackData::MODE1:
    case TrackData::MODE2_FORM1:
      out << "MODE1/2048";
      break;
    case TrackData::MODE2:
    case TrackData::MODE2_FORM_MIX:
      out << "MODE2/2336";
      break;
    case TrackData::MODE1_RAW:
      out << "MODE1/2352";
      break;
    case TrackData::MODE2_RAW:
      out << "MODE2/2352";
      break;
    default:
      break;
    }
    
    out << "\n";

    if ( trun->copyPermitted() ) {
      out << "    FLAGS DCP\n";
    }

    const SubTrack *strun;
    SubTrackIterator stitr(trun);
    int pregap = 0;

    for (strun = stitr.first(); strun != NULL; strun = stitr.next()) {
      if (strun->TrackData::type() == TrackData::ZERODATA) {
	out << "    PREGAP " << trun->start().str() << "\n";
	pregap = 1;
      }
      else {
	if (!pregap && trun->start().lba() != 0) {
	  out << "    INDEX 00 " << Msf(offset).str() << "\n";
	  out << "    INDEX 01 " 
	      << Msf(offset + trun->start().lba()).str() << "\n";
	}
	else {
	  out << "    INDEX 01 " << Msf(offset).str() << "\n";
	}

	offset += trun->length().lba();

	if (pregap)
	  offset -= trun->start().lba();
      }
    }
  } 

  out.close();

  message(1, "Converted toc-file '%s' to cue file '%s'.",
	  tocFile, cueFile);
  message(1, "");
  message(1, "Please note that the resulting cue file is only valid if the");
  message(1, "toc-file was created with cdrdao using the commands 'read-toc'");
  message(1, "or 'read-cd'. For manually created or edited toc-files the");
  message(1, "cue file may not be correct. This program just checks for");
  message(1, "the most obvious toc-file features that cannot be converted to");
  message(1, "a cue file.");
  message(1, "Furthermore, if the toc-file contains audio tracks the byte");
  message(1, "order of the image file will be wrong which results in static");
  message(1, "noise when the resulting cue file is used for recording");
  message(1, "(even with cdrdao itself).");

  return 0;
}
